src/snowflake/cli/plugins/nativeapp/artifacts.py

```diff
@@ -101,6 +101,109 @@ def __init__(
+        # built-in dict instances are ordered as of Python 3.7
+        self.__src_to_dest: Dict[Path, List[Path]] = {}
+        self.__dest_to_src: Dict[Path, List[Path]] = {}
+        self._dest_is_dir: Dict[Path, bool] = {}
+
+    def put(self, src: Path, dest: Path, dest_is_dir: bool) -> None:
+        """
+        Adds a new source-destination mapping pair to this map, if necessary.
+
+        Arguments:
+            src {Path} -- the source path, in canonical form.
+            dest {Path} -- the destination path, in canonical form.
+            dest_is_dir {bool} -- whether the destination path is a directory.
+        """
+        # Both paths should be in canonical form
+        assert not src.is_absolute()
+        assert not dest.is_absolute()
+
+        absolute_src = self._project_root / src
+
+        current_sources = self.__dest_to_src.get(dest, [])
+        src_is_dir = absolute_src.is_dir()
+        if dest_is_dir:
+            # directory -> directory
+            # Check that dest is currently unmapped
+            current_is_dir = self._dest_is_dir.get(dest, False)
+            if current_is_dir:
+                # mapping to an existing directory is not allowed
+                raise TooManyFilesError(dest)
+        else:
+            # file -> file
+            # Check that there is no previous mapping for the same file.
+            if current_sources and src not in current_sources:
+                # There is already a different source mapping to this destination
+                raise TooManyFilesError(dest)
+
+        if src_is_dir:
+            # mark all subdirectories of this source as directories so that we can
+            # detect accidental clobbering
+            for (root, _, files) in os.walk(absolute_src, followlinks=True):
+                canonical_subdir = Path(root).relative_to(absolute_src)
+                canonical_dest_subdir = dest / canonical_subdir
+                self._update_dest_is_dir(canonical_dest_subdir, is_dir=True)
+                for f in files:
+                    self._update_dest_is_dir(canonical_dest_subdir / f, is_dir=False)
+
+        # make sure we check of dest_is_dir consistency regardless of whether the
+        # insertion happened. This update can fail, so we need to do it first to
+        # avoid applying partial updates to the underlying data storage.
+        self._update_dest_is_dir(dest, dest_is_dir)
+
+        dests = self.__src_to_dest.setdefault(src, [])
+        srcs = self.__dest_to_src.setdefault(dest, [])
+        if dest not in dests:
+            dests.append(dest)
+            srcs.append(src)
+            self.__src_dest_pairs.append((src, dest))
+
+    def get_sources(self, dest: Path) -> Iterable[Path]:
+        """
+        Returns all source paths associated with the provided destination path, in insertion order.
+        """
+        return self.__dest_to_src.get(dest, [])
+
+    def get_destinations(self, src: Path) -> Iterable[Path]:
```

Comment:
For 181 and 175, the type for values in each of these dicts is defined as `List[Path]` instead of more generic `Iterable`. Do we expect any other type of `Iterable` to be returned from these?

Reply:
If I return a list, I should make it immutably or make a copy of it so it doesn't get modified outside of this class. I didn't think the overhead of the copy was justified, so I went for Iterable instead to say "you can look but not modify".

Reply:
Okay, I did not know `Iterable` also does that too, was only aware of `Tuple` or `FrozenSet` offering immutability.

Reply:
it doesn't _do_ anything to be clear. It's just giving any human (or IDE) a hint that the value returned shouldn't be used as a mutable list (although it is, and could be mistakenly used that way). Since we control the entire logic, it felt OK to me.